src/app/AttributeCache.cpp

```diff
@@ -219,5 +239,113 @@ CHIP_ERROR AttributeCache::GetStatus(const ConcreteAttributePath & path, StatusI
     return CHIP_NO_ERROR;
 }
 
+void AttributeCache::UpdateFilterMap(std::map<DataVersionFilter, size_t> & aMap)
+{
+    for (auto const & endpointIter : mCache)
+    {
+        EndpointId endpointId = endpointIter.first;
+        for (auto const & clusterIter : endpointIter.second)
+        {
+            DataVersion dataVersion = 0;
+            uint32_t clusterSize    = 0;
+            ClusterId clusterId     = clusterIter.first;
+            for (auto const & attributeIter : clusterIter.second)
+            {
+                if (!attributeIter.second.Is<StatusIB>())
+                {
+                    TLV::TLVReader reader;
+                    System::PacketBufferTLVReader bufReader;
+                    bufReader.Init(attributeIter.second.Get<VersionedAttributeBuffer>().mHandle.Retain());
+                    ReturnOnFailure(bufReader.Next());
+
+                    reader.Init(bufReader);
+                    // Skip to the end of the element.
+                    ReturnOnFailure(reader.Skip());
+
+                    // Compute the amount of value data
+                    clusterSize += reader.GetLengthRead();
+
+                    dataVersion = attributeIter.second.Get<VersionedAttributeBuffer>().mDataVersion;
+                }
+            }
+            if (clusterSize == 0)
+            {
+                continue;
+            }
+            DataVersionFilter filter(endpointId, clusterId, dataVersion);
+            aMap[filter] = clusterSize;
```

Comment:
Had a discussion with @yunhanw-google on some of the issues here:

In the case where we have multiple attributes in the cache (for a given cluster instance) at different versions, it's going to result in multiple data version filter entries in the map that point to the same cluster, but have different data versions. This in-and-of-itself isn't a problem.

While this is handled [correctly](https://github.com/project-chip/connectedhomeip/blob/master/src/app/reporting/Engine.cpp#L66) on the server, it causes a different problem: when this cache is used with wildcard subscriptions that encompass the entirety of a cluster or wider, over-time, the variance of data versions for attributes in a given cluster is going to keep increasing (i.e the set of data versions for a given cluster will keep increasing till its hits the limit of # of attributes in the cluster). This will result in the 'duplication index' of the data version filter list increasing as well (i.e more and more entries that refer to the same cluster but have different data version values).

This actually will severely degrade the utility of data version filters, since not only will that list not fit in an IPv6 MTU, it will hit the limits of what the server can actually process given the bounded # of filter paths it has. And this degradation happens over-time as the cache gets fragmented.

My original suggestion of shifting where we store the data version in the cache away from per/attribute to be per/cluster would solve that. 

However, that brings about a different issue: if you try to use the cache across different dis-joint reads/subscribes where each interaction specifies a disjoint set of paths to attributes, it is possible for the cache to contain attributes whose value is actually stale relative to the data version stored. This is actually 'incoherent' and incorrect, and can cause trouble. One alternative would to be say that applications shouldn't use the cache in this manner, but I think we shouldn't actually limit these if possible, since the actual data in the cache is 'up to date' relative to what the client/application desires for the set of reads they are doing, so all we really want to do is to just infer that the data version stored is not coherent/accurate and use that as a signal to avoid adding an entry to this data version filter list.

That coherence inference is quite simple - if we encounter attributes in a given cluster instance that are not all of the same data version, we infer that there is no coherence and if so, avoid adding an entry to the map.

Consequently, we should also make insertion into the map unique to just the `ClusterPath` and not the tuple of path + data version.

With those changes, we should be good to go.



Phew. That was a tough one...

Reply:
Another issue we have not yet settle down is regarding the below situation, if we read C1A1, then read C1 with cached C1 data version filter, then second read cannot retrieve all attributes for C1.

Reply:
another situation is that, if we read C1A1, then read C1A2 with cached C1 data version filter from first read, then second read would not get anything..

Reply:
it seems like we should only allow wildcard AttributePath to use AttributeCache feature with DataVersion, which can bypass the above the situations.

Reply:
Good catch @yunhanw-google.

I think the only way to solve this (and the previous cases) is to do one of the two following options:

1. Do not re-use data versions across different `ReadClient` instances. Interactions within a specific `ReadClient` instance will have a fixed path set, so we leverage that to correctly populate the data versions on subsequent interactions on the same instance, which only happen for re-subscriptions. The con with this approach is that we cannot use the cache across Reads.

Each instance would need some kind of unique instance ID, which we can mint statically in the IM engine (or some equivalent).

2. Track the path set associated with each Read/Subscribe request in the cache. Only re-use the data versions for a given cluster instance if the following is true: 

For each cluster in the cache, track the set of paths in the request that affect that cluster instance. This is the 'existing' path set (that accompanies the data version embedded in the cache).

For each path in the 'new' path-set associated with a new interaction, assess if the following are true when compared with every path in the existing set for each cluster instance:
     1. The 'new' path is a strict sub-set of the existing path.
     2. The existing path is wildcard of the cluster (or broader).

If the above two are are not true for any of the comparisons, the data version shall not be re-used.

This is more complicated to implement, but can be re-used for reads.